e2e_playwright/st_fragments_nested_test.py

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.shared.app_utils import click_button
+
+
+def get_uuids(app: Page):
+    expect(app.get_by_test_id("stMarkdown")).to_have_count(3)
+
+    outside_fragment_text = app.get_by_test_id("stMarkdown").first.text_content()
+    inner_fragment_text = app.get_by_test_id("stMarkdown").nth(1).text_content()
+    outer_fragment_text = app.get_by_test_id("stMarkdown").last.text_content()
+
+    return outside_fragment_text, inner_fragment_text, outer_fragment_text
+
+
+def test_full_app_rerun(app: Page):
+    outside_fragment_text, inner_fragment_text, outer_fragment_text = get_uuids(app)
+
+    click_button(app, "rerun whole app")
+
+    expect(app.get_by_test_id("stMarkdown").first).not_to_have_text(
+        outside_fragment_text
+    )
+    expect(app.get_by_test_id("stMarkdown").nth(1)).not_to_have_text(
+        inner_fragment_text
+    )
+    expect(app.get_by_test_id("stMarkdown").last).not_to_have_text(outer_fragment_text)
```

Comment:
Very nice way to check which parts of the app ran! 🧠 

nit: it took me a moment to understand that the app uses `uuid`s  that change when you click one of the buttons, so that the uuids change depending on full app run, out fragment etc. It might make sense to add a quick comment ala

```python
"""When we click one of the buttons, the respective uuids change depending on which parts of the app run. So we know which parts ran by comparing the uuids."""
```

Additionally, it might be elegant to change the tests to this pattern:

```python
outside_fragment_text, inner_fragment_text, outer_fragment_text = get_uuids(app)

click_button(app, "rerun whole app")

new_outside_fragment_text, new_inner_fragment_text, new_outer_fragment_text = get_uuids(app)

expect(new_outside_fragment_text).not_to_have_text(outside_fragment_text)
expect(new_inner_fragment_text).not_to_have_text(inner_fragment_text)
expect(new_outer_fragment_text).not_to_have_text(outer_fragment_text)
```
same for the other tests.
If this is possible, it might be easier to see what is expected to change etc 🙂

Reply:
👍 Added some explanatory comments.

I believe the way that the `expect` calls after the app is rerun were written is intentional, though, because otherwise there's a chance for a race to occur where the `expect` calls are made too quickly (since they'd be done against strings rather than whatever playwright uses to wrap DOM elements), and an error is thrown even though the UUID displayed in the app will soon update.